06_puppy/alextmz/mapstore.go

```diff
@@ -0,0 +1,57 @@
+package main
+
+import (
+	"fmt"
+)
+
+type MapStore map[int]Puppy
+
+func NewmapStore() MapStore {
+	a := MapStore{}
+	return a
+}
+
+func (m MapStore) CreatePuppy(p *Puppy) error {
+	switch {
+	case p == nil:
+		return fmt.Errorf("error, puppy pointer is nil")
+	case p.ID == 0:
+		p.ID = len(m) + 1
+		m[p.ID] = *p
+		return nil
+	default:
+		return fmt.Errorf("trying to create a puppy that already has an ID %d", p.ID)
+	}
```

Comment:
no reference found, but i'd typically write this as two ifs:
```go
if p == nil {
    return fmt.Errof("...")
}
if p.ID != 0 {
    return fmt.Errof("...")
}
p.ID = len(m) + 1
m[p.ID] = *p
return nil

```

Reply:
Yes, I wrote if's but then I stumbled on https://tour.golang.org/flowcontrol/11 and it just seemed more straightforward to read and to see the code flow. 
Specifically besides the use of switch/if, I used the cases in this order as reading it goes like "nil is an error, zero is okay, and everything else is error" - nothing else to think about, you know all the error and ok conditions of flow. 
Using your test ordering one would have to read it as "nil is error, not zero is error, and everything else is okay" - leaving to the programmer to deduce further that the only okay value is zero through a double-inversion, which is straightforward to be fair, but more work than not thinking at all as in the former code, as it's explicit. 
Just seems more intuitive and I'd like to keep it that way if the language standards don't have a problem with it :-) 
Would be really curious to hear from more people about this if my reasoning makes any sense or it's some kind of beginner mistake and I don't see it, which is highly probable...

Reply:
Changed my mind, to changing this to `if`, due to:
1. Use of switches like that is useful but gain is marginal with that small of a function.
2. Rest of the code uses `if` and it's going to look odd. 

Hah, btw found a reference: https://golang.org/doc/effective_go.html#switch

Reply:
awesome, thanks for the link to the reference.